File: middleware/appAuth.js
const jwt = require('jsonwebtoken');
const { AppUser, User } = require('../models');

/**
 * Authentication middleware for AppUsers
 * Verifies JWT token and attaches app user to context
 */
const authenticateApp = async (ctx, next) => {
  try {
    // Get token from header
    const authHeader = ctx.headers.authorization;

    if (!authHeader) {
      ctx.status = 401;
      ctx.body = { error: 'No authorization header provided' };
      return;
    }

    // Extract token (Bearer <token>)
    const token = authHeader.startsWith('Bearer ')
      ? authHeader.substring(7)
      : authHeader;

    if (!token) {
      ctx.status = 401;
      ctx.body = { error: 'No token provided' };
      return;
    }

    // Verify token
    const decoded = jwt.verify(token, global.cfg.jwt.secret);

    // Ensure this is an app user token, not an admin token
    if (decoded.type !== 'appUser') {
      ctx.status = 401;
      ctx.body = { error: 'Invalid token type' };
      return;
    }

    // Get app user from database
    const appUser = await AppUser.findByPk(decoded.id);

    if (!appUser) {
      ctx.status = 401;
      ctx.body = { error: 'User not found' };
      return;
    }

    if (!appUser.isActive) {
      ctx.status = 401;
      ctx.body = { error: 'User account is deactivated' };
      return;
    }

    // Attach app user to context
    ctx.state.appUser = appUser;

    await next();
  } catch (error) {
    if (error.name === 'JsonWebTokenError') {
      ctx.status = 401;
      ctx.body = { error: 'Invalid token' };
    } else if (error.name === 'TokenExpiredError') {
      ctx.status = 401;
      ctx.body = { error: 'Token expired' };
    } else {
      ctx.status = 500;
      ctx.body = { error: 'Authentication error' };
    }
  }
};

/**
 * Accepts either an app user JWT or a management user JWT.
 * Used for routes that both the Flutter app and the management UI need to access.
 */
const authenticateAppOrAdmin = async (ctx, next) => {
  const authHeader = ctx.headers.authorization;
  if (!authHeader) {
    ctx.status = 401;
    ctx.body = { error: 'No authorization header provided' };
    return;
  }

  const token = authHeader.startsWith('Bearer ')
    ? authHeader.substring(7)
    : authHeader;

  if (!token) {
    ctx.status = 401;
    ctx.body = { error: 'No token provided' };
    return;
  }

  try {
    const decoded = jwt.verify(token, global.cfg.jwt.secret);

    if (decoded.type === 'appUser') {
      const appUser = await AppUser.findByPk(decoded.id);
      if (!appUser || !appUser.isActive) {
        ctx.status = 401;
        ctx.body = { error: 'User not found or inactive' };
        return;
      }
      ctx.state.appUser = appUser;
    } else {
      const user = await User.findByPk(decoded.id);
      if (!user || !user.isActive) {
        ctx.status = 401;
        ctx.body = { error: 'User not found or inactive' };
        return;
      }
      ctx.state.user = user;
    }

    await next();
  } catch (error) {
    if (error.name === 'JsonWebTokenError') {
      ctx.status = 401;
      ctx.body = { error: 'Invalid token' };
    } else if (error.name === 'TokenExpiredError') {
      ctx.status = 401;
      ctx.body = { error: 'Token expired' };
    } else {
      ctx.status = 500;
      ctx.body = { error: 'Authentication error' };
    }
  }
};

module.exports = {
  authenticateApp,
  authenticateAppOrAdmin,
};

File: backend_go/internal/middleware/app_auth.go
package middleware

import (
	"household-go/internal/database"
	"household-go/internal/models"
	"household-go/internal/utils"

	"github.com/gin-gonic/gin"
)

// AuthenticateApp verifies the JWT and attaches the AppUser to the Gin context.
// Equivalent to your middleware/appAuth.js `authenticateApp` function.
//
// Key difference from Authenticate: we check claims.Type == "appUser" to prevent
// admin tokens from being used on app endpoints and vice versa.
func AuthenticateApp(cfg AuthConfig) gin.HandlerFunc {
	return func(c *gin.Context) {
		token := extractBearerToken(c)
		if token == "" {
			utils.Unauthorized(c, "No authorization token provided")
			c.Abort()
			return
		}

		claims, err := utils.ParseToken(token, cfg.JWTSecret)
		if err != nil {
			utils.Unauthorized(c, "Invalid or expired token")
			c.Abort()
			return
		}

		// Token-type guard: ensures app tokens can't access admin routes and vice versa.
		if claims.Type != "appUser" {
			utils.Unauthorized(c, "Invalid token type")
			c.Abort()
			return
		}

		var appUser models.AppUser
		if err := database.DB.First(&appUser, claims.ID).Error; err != nil {
			utils.Unauthorized(c, "User not found")
			c.Abort()
			return
		}

		if !appUser.IsActive {
			utils.Unauthorized(c, "Account is deactivated")
			c.Abort()
			return
		}

		// Store as "appUser" key — different from "authUser" so both can coexist
		// in a request that's been authenticated by both middleware layers (unlikely
		// but consistent with the two-token architecture).
		c.Set("appUser", &appUser)
		c.Next()
	}
}
